File: confusion_engine.py
"""
Color Confusion — Python Game Engine
Stroop Effect question generation, validation, and scoring logic.
Used by the Flask API to serve game data to the frontend.
"""

import random
import time
import json
from dataclasses import dataclass, field, asdict
from typing import List, Dict, Optional, Tuple

# ── Color Palette ─────────────────────────────────────────────
COLORS = {
    'Red':      '#ef4444',
    'Blue':     '#3b82f6',
    'Green':    '#22c55e',
    'Yellow':   '#eab308',
    'Purple':   '#8b5cf6',
    'Orange':   '#f97316',
    'Pink':     '#ff29ff',
    'Cyan':     '#06b6d4',
    'Indigo':   '#6366f1',
    'Violet':   '#8b5cf6',
    'Black':    '#1a1a1a',
    'Brown':    '#78350f',
    'Lavender': '#a78bfa',
    'White':    '#ffffff',
    'Beige':    '#f5f5dc',
}

# Difficulty tiers — more colors = harder
DIFFICULTY_TIERS = {
    1: ['Red', 'Blue', 'Green', 'Yellow'],
    2: ['Red', 'Blue', 'Green', 'Yellow', 'Purple', 'Orange'],
    3: ['Red', 'Blue', 'Green', 'Yellow', 'Purple', 'Orange', 'Pink', 'Cyan'],
    4: ['Red', 'Blue', 'Green', 'Yellow', 'Purple', 'Orange', 'Pink', 'Cyan', 'Indigo', 'Violet'],
    5: list(COLORS.keys()),  # All colors
}


@dataclass
class StroopQuestion:
    """A single Stroop effect question."""
    text_word: str          # The word displayed (e.g., "YELLOW")
    font_color_name: str    # The actual font color name (e.g., "Red")
    font_color_hex: str     # Hex for the font color
    options: List[str]      # 4 answer choices
    difficulty: int         # Current difficulty level
    timestamp: float = field(default_factory=time.time)

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class AnswerResult:
    """Result of a single answer validation."""
    correct: bool
    reaction_time_ms: int
    points_earned: int
    combo: int
    speed_bonus: int
    multiplier: float


class ConfusionEngine:
    """
    Core Stroop Effect game engine.
    
    Handles question generation, answer validation,
    scoring with combo multipliers, and difficulty scaling.
    """
    
    def __init__(self, difficulty: int = 1):
        self.difficulty = max(1, min(5, difficulty))
        self._color_pool = self._get_color_pool()
    
    def _get_color_pool(self) -> List[str]:
        """Get available colors based on difficulty."""
        tier = min(self.difficulty, max(DIFFICULTY_TIERS.keys()))
        return DIFFICULTY_TIERS.get(tier, DIFFICULTY_TIERS[1])
    
    def generate_question(self) -> StroopQuestion:
        """
        Generate a Stroop question where the displayed word
        and its font color are always different (the Stroop effect).
        """
        pool = self._color_pool
        
        # Pick the font color (this is the CORRECT answer)
        font_color_name = random.choice(pool)
        
        # Pick a DIFFERENT word to display (creates the Stroop effect)
        available_words = [c for c in pool if c != font_color_name]
        text_word = random.choice(available_words) if available_words else 'Black'
        
        # Build 4 options: correct + 3 distractors
        distractors = [c for c in pool if c != font_color_name and c != text_word]
        random.shuffle(distractors)
        options = [font_color_name] + distractors[:3]
        
        # Ensure we have exactly 4 options
        while len(options) < 4:
            filler = random.choice(list(COLORS.keys()))
            if filler not in options:
                options.append(filler)
        
        random.shuffle(options)
        
        return StroopQuestion(
            text_word=text_word.upper(),
            font_color_name=font_color_name,
            font_color_hex=COLORS.get(font_color_name, '#888'),
            options=options,
            difficulty=self.difficulty
        )
    
    def validate_answer(
        self,
        question: StroopQuestion,
        selected_color: str,
        reaction_time_ms: int,
        current_combo: int
    ) -> AnswerResult:
        """
        Validate a player's answer and calculate points.
        
        Scoring:
        - Base: 10 points per correct answer
        - Speed bonus: up to 20 extra points for fast reactions (<2s)
        - Combo multiplier: 1 + (combo * 0.1), e.g., 5-streak = 1.5x
        """
        correct = selected_color.lower() == question.font_color_name.lower()
        
        if correct:
            combo = current_combo + 1
            base_points = 10
            speed_bonus = max(0, (2000 - reaction_time_ms) // 100)
            multiplier = 1.0 + (combo * 0.1)
            points = round((base_points + speed_bonus) * multiplier)
        else:
            combo = 0
            base_points = 0
            speed_bonus = 0
            multiplier = 1.0
            points = 0
        
        return AnswerResult(
            correct=correct,
            reaction_time_ms=reaction_time_ms,
            points_earned=points,
            combo=combo,
            speed_bonus=speed_bonus,
            multiplier=multiplier
        )
    
    def scale_difficulty(self, score: int) -> None:
        """Auto-scale difficulty based on score milestones."""
        if score >= 40:
            self.difficulty = 5
        elif score >= 30:
            self.difficulty = 4
        elif score >= 20:
            self.difficulty = 3
        elif score >= 10:
            self.difficulty = 2
        else:
            self.difficulty = 1
        self._color_pool = self._get_color_pool()
    
    def get_performance_rating(
        self,
        avg_reaction_ms: float,
        total_score: int
    ) -> str:
        """Assign a performance rating based on reaction time and score."""
        if avg_reaction_ms < 600 and total_score > 40:
            return "Grandmaster"
        elif avg_reaction_ms < 800 and total_score > 25:
            return "Expert"
        elif avg_reaction_ms < 1000 and total_score > 15:
            return "Advanced"
        elif avg_reaction_ms < 1200 and total_score > 8:
            return "Intermediate"
        elif total_score > 3:
            return "Beginner"
        else:
            return "Trainee"


# ── Session Manager ───────────────────────────────────────────

class GameSession:
    """
    Manages a complete Color Confusion game session.
    
    Modes:
    - endless:   3 lives, play until lives run out
    - survival:  60s timer, +3s correct / -3s wrong
    - speed:     race to 50 correct answers
    """
    
    def __init__(self, mode: str = 'endless'):
        self.mode = mode
        self.engine = ConfusionEngine(difficulty=1)
        self.score = 0
        self.total_points = 0
        self.combo = 0
        self.max_combo = 0
        self.lives = 3 if mode == 'endless' else -1
        self.time_left = 60.0 if mode == 'survival' else -1
        self.target = 50 if mode == 'speed' else -1
        self.reactions: List[int] = []
        self.start_time = time.time()
        self.is_active = True
        self.current_question: Optional[StroopQuestion] = None
    
    def next_question(self) -> Optional[StroopQuestion]:
        """Generate the next question if session is still active."""
        if not self.is_active:
            return None
        self.current_question = self.engine.generate_question()
        return self.current_question
    
    def submit_answer(self, selected_color: str, reaction_time_ms: int) -> dict:
        """Process an answer and return the result with updated session state."""
        if not self.is_active or not self.current_question:
            return {"error": "No active question"}
        
        result = self.engine.validate_answer(
            self.current_question, selected_color, reaction_time_ms, self.combo
        )
        
        self.reactions.append(reaction_time_ms)
        
        if result.correct:
            self.score += 1
            self.combo = result.combo
            self.max_combo = max(self.max_combo, self.combo)
            self.total_points += result.points_earned
            
            # Mode-specific rewards
            if self.mode == 'survival':
                self.time_left += 3
            
            # Difficulty scaling every 5 correct
            if self.score % 5 == 0:
                self.engine.scale_difficulty(self.score)
        else:
            self.combo = 0
            
            # Mode-specific penalties
            if self.mode == 'endless':
                self.lives -= 1
                if self.lives <= 0:
                    self.is_active = False
            elif self.mode == 'survival':
                self.time_left = max(0, self.time_left - 3)
                if self.time_left <= 0:
                    self.is_active = False
            elif self.mode == 'speed':
                self.total_points = max(0, self.total_points - 5)
        
        # Speed run win condition
        if self.mode == 'speed' and self.score >= self.target:
            self.is_active = False
        
        return {
            "correct": result.correct,
            "points_earned": result.points_earned,
            "total_points": self.total_points,
            "score": self.score,
            "combo": self.combo,
            "max_combo": self.max_combo,
            "lives": self.lives,
            "time_left": round(self.time_left, 1) if self.time_left >= 0 else -1,
            "is_active": self.is_active,
            "speed_bonus": result.speed_bonus,
            "multiplier": result.multiplier,
        }
    
    def get_final_report(self) -> dict:
        """Generate the end-of-session report."""
        elapsed = round(time.time() - self.start_time, 2)
        avg_rt = (
            round(sum(self.reactions) / len(self.reactions))
            if self.reactions else 0
        )
        rating = self.engine.get_performance_rating(avg_rt, self.score)
        
        return {
            "mode": self.mode,
            "total_points": self.total_points,
            "score": self.score,
            "max_combo": self.max_combo,
            "avg_reaction_ms": avg_rt,
            "elapsed_seconds": elapsed,
            "rating": rating,
            "total_questions": len(self.reactions),
            "accuracy": round(self.score / max(1, len(self.reactions)) * 100, 1),
        }


# ── CLI Test ──────────────────────────────────────────────────
if __name__ == '__main__':
    print("=== Color Confusion Engine Test ===\n")
    
    engine = ConfusionEngine(difficulty=2)
    
    for i in range(5):
        q = engine.generate_question()
        print(f"Q{i+1}: Word='{q.text_word}' | Font Color='{q.font_color_name}' ({q.font_color_hex})")
        print(f"     Options: {q.options}")
        
        # Simulate correct answer
        result = engine.validate_answer(q, q.font_color_name, random.randint(300, 1500), i)
        print(f"     -> Points: {result.points_earned} | Combo: {result.combo} | Multiplier: {result.multiplier}x")
        print()
    
    print("--- Session Test (Endless) ---")
    session = GameSession('endless')
    for i in range(8):
        q = session.next_question()
        if not q:
            break
        # Alternate correct/wrong for testing
        answer = q.font_color_name if i % 3 != 0 else 'WrongColor'
        result = session.submit_answer(answer, random.randint(200, 2000))
        status = 'OK' if result['correct'] else 'X'
        print(f"  Round {i+1}: {status} | "
              f"Score:{result['score']} | Points:{result['total_points']} | "
              f"Lives:{result['lives']} | Combo:{result['combo']}")
    
    report = session.get_final_report()
    print(f"\nFinal Report: {json.dumps(report, indent=2)}")

File: ConfusionEngine.cs
/**
 * Color Confusion — C# Game Session Manager
 * 
 * State machine for managing Endless, Survival, and Speed Run modes.
 * Centralized scoring, session lifecycle, and result generation.
 * 
 * Build:  dotnet build  (or csc ConfusionEngine.cs)
 * Run:    dotnet run     (or ./ConfusionEngine)
 */

using System;
using System.Collections.Generic;
using System.Linq;

namespace ColorConfusion
{
    // ── Color Registry ───────────────────────────────────────

    public static class ColorRegistry
    {
        public static readonly Dictionary<string, string> Colors = new()
        {
            { "Red",      "#ef4444" },
            { "Blue",     "#3b82f6" },
            { "Green",    "#22c55e" },
            { "Yellow",   "#eab308" },
            { "Purple",   "#8b5cf6" },
            { "Orange",   "#f97316" },
            { "Pink",     "#ff29ff" },
            { "Cyan",     "#06b6d4" },
            { "Indigo",   "#6366f1" },
            { "Violet",   "#8b5cf6" },
            { "Black",    "#1a1a1a" },
            { "Brown",    "#78350f" },
            { "Lavender", "#a78bfa" },
            { "White",    "#ffffff" },
            { "Beige",    "#f5f5dc" },
        };

        /// <summary>
        /// Returns the color pool for a given difficulty level (1-5).
        /// Higher difficulty = more colors to choose from.
        /// </summary>
        public static List<string> GetColorPool(int difficulty)
        {
            int count = Math.Min(Colors.Count, 4 + (difficulty - 1) * 2);
            return Colors.Keys.Take(count).ToList();
        }
    }

    // ── Game Mode Enum ───────────────────────────────────────

    public enum GameMode
    {
        Endless,    // 3 lives, play until out
        Survival,   // 60s timer, +3s correct / -3s wrong
        SpeedRun    // Race to 50 correct answers
    }

    // ── Game State Enum ──────────────────────────────────────

    public enum GameState
    {
        Idle,       // Waiting to start
        Active,     // Playing
        Paused,     // Paused (future use)
        Finished    // Game over
    }

    // ── Stroop Question ──────────────────────────────────────

    public class StroopQuestion
    {
        public string TextWord { get; set; } = "";
        public string FontColorName { get; set; } = "";
        public string FontColorHex { get; set; } = "";
        public List<string> Options { get; set; } = new();

        public override string ToString() =>
            $"Word='{TextWord}' FontColor='{FontColorName}' Options=[{string.Join(", ", Options)}]";
    }

    // ── Score Calculator ─────────────────────────────────────

    public static class ScoreCalculator
    {
        /// <summary>
        /// Calculate points for a correct answer with combo and speed bonuses.
        /// </summary>
        public static int CalculatePoints(int reactionTimeMs, int currentCombo, int difficulty)
        {
            int basePoints = 10;
            int speedBonus = Math.Max(0, (2000 - reactionTimeMs) / 100);
            double comboMultiplier = 1.0 + (currentCombo * 0.1);
            double difficultyBonus = 1.0 + (difficulty - 1) * 0.15;

            return (int)Math.Round((basePoints + speedBonus) * comboMultiplier * difficultyBonus);
        }

        /// <summary>
        /// Calculate coins earned (1 per 100 points).
        /// </summary>
        public static int CalculateCoins(int totalPoints) => totalPoints / 100;

        /// <summary>
        /// Calculate stars earned (1 per 10 correct answers).
        /// </summary>
        public static int CalculateStars(int correctAnswers) => correctAnswers / 10;
    }

    // ── Session Result ───────────────────────────────────────

    public class SessionResult
    {
        public GameMode Mode { get; set; }
        public int TotalPoints { get; set; }
        public int CorrectAnswers { get; set; }
        public int WrongAnswers { get; set; }
        public int MaxCombo { get; set; }
        public double AvgReactionMs { get; set; }
        public double FastestReaction { get; set; }
        public double SlowestReaction { get; set; }
        public double ElapsedSeconds { get; set; }
        public double Accuracy { get; set; }
        public string Rating { get; set; } = "Trainee";
        public int CoinsEarned { get; set; }
        public int StarsEarned { get; set; }

        /// <summary>
        /// Assign performance rating based on Stroop metrics.
        /// </summary>
        public void CalculateRating()
        {
            if (AvgReactionMs < 400 && CorrectAnswers > 50)       Rating = "Legendary";
            else if (AvgReactionMs < 500 && CorrectAnswers > 40)  Rating = "Grandmaster";
            else if (AvgReactionMs < 600 && CorrectAnswers > 30)  Rating = "Master";
            else if (AvgReactionMs < 700 && CorrectAnswers > 25)  Rating = "Expert";
            else if (AvgReactionMs < 800 && CorrectAnswers > 20)  Rating = "Advanced";
            else if (AvgReactionMs < 1000 && CorrectAnswers > 15) Rating = "Proficient";
            else if (AvgReactionMs < 1200 && CorrectAnswers > 10) Rating = "Intermediate";
            else if (AvgReactionMs < 1500 && CorrectAnswers > 5)  Rating = "Beginner";
            else Rating = "Trainee";
        }

        public override string ToString()
        {
            return $@"
╔══════════════════════════════════════════╗
║     COLOR CONFUSION — SESSION REPORT     ║
╠══════════════════════════════════════════╣
║  Mode:           {Mode,-24}║
║  Total Points:   {TotalPoints,-24}║
║  Correct:        {CorrectAnswers,-24}║
║  Wrong:          {WrongAnswers,-24}║
║  Accuracy:       {Accuracy:F1}%{new string(' ', 20 - Accuracy.ToString("F1").Length)}║
║  Max Combo:      {MaxCombo,-24}║
║  Avg Reaction:   {AvgReactionMs:F0}ms{new string(' ', 21 - AvgReactionMs.ToString("F0").Length)}║
║  Fastest:        {FastestReaction:F0}ms{new string(' ', 21 - FastestReaction.ToString("F0").Length)}║
║  Slowest:        {SlowestReaction:F0}ms{new string(' ', 21 - SlowestReaction.ToString("F0").Length)}║
║  Elapsed:        {ElapsedSeconds:F2}s{new string(' ', 22 - ElapsedSeconds.ToString("F2").Length)}║
║  Rating:         {Rating,-24}║
║  Coins Earned:   {CoinsEarned,-24}║
║  Stars Earned:   {StarsEarned,-24}║
╚══════════════════════════════════════════╝";
        }
    }

    // ── Game Session — State Machine ─────────────────────────

    public class GameSession
    {
        private readonly Random _random = new();
        private readonly List<long> _reactions = new();
        private List<string> _colorPool;

        // Session state
        public GameMode Mode { get; }
        public GameState State { get; private set; } = GameState.Idle;
        public int Score { get; private set; } = 0;
        public int TotalPoints { get; private set; } = 0;
        public int Combo { get; private set; } = 0;
        public int MaxCombo { get; private set; } = 0;
        public int Lives { get; private set; }
        public double TimeLeft { get; private set; }
        public int Target { get; private set; }
        public int Difficulty { get; private set; } = 1;
        public int WrongAnswers { get; private set; } = 0;
        public StroopQuestion? CurrentQuestion { get; private set; }
        public DateTime StartTime { get; private set; }

        public GameSession(GameMode mode)
        {
            Mode = mode;
            Lives = mode == GameMode.Endless ? 3 : -1;
            TimeLeft = mode == GameMode.Survival ? 60.0 : -1;
            Target = mode == GameMode.SpeedRun ? 50 : -1;
            _colorPool = ColorRegistry.GetColorPool(Difficulty);
        }

        /// <summary>Start the game session.</summary>
        public void Start()
        {
            State = GameState.Active;
            StartTime = DateTime.Now;
        }

        /// <summary>Generate the next Stroop question.</summary>
        public StroopQuestion? NextQuestion()
        {
            if (State != GameState.Active) return null;

            string fontColor = _colorPool[_random.Next(_colorPool.Count)];
            var wordCandidates = _colorPool.Where(c => c != fontColor).ToList();
            string textWord = wordCandidates.Count > 0
                ? wordCandidates[_random.Next(wordCandidates.Count)]
                : "Black";

            // Build options
            var distractors = _colorPool.Where(c => c != fontColor && c != textWord).ToList();
            distractors = distractors.OrderBy(_ => _random.Next()).ToList();

            var options = new List<string> { fontColor };
            options.AddRange(distractors.Take(3));

            while (options.Count < 4)
            {
                var filler = ColorRegistry.Colors.Keys.ElementAt(
                    _random.Next(ColorRegistry.Colors.Count));
                if (!options.Contains(filler)) options.Add(filler);
            }
            options = options.OrderBy(_ => _random.Next()).ToList();

            CurrentQuestion = new StroopQuestion
            {
                TextWord = textWord.ToUpper(),
                FontColorName = fontColor,
                FontColorHex = ColorRegistry.Colors.GetValueOrDefault(fontColor, "#888"),
                Options = options
            };

            return CurrentQuestion;
        }

        /// <summary>Submit an answer and get the result.</summary>
        public Dictionary<string, object> SubmitAnswer(string selectedColor, int reactionTimeMs)
        {
            if (State != GameState.Active || CurrentQuestion == null)
                return new Dictionary<string, object> { { "error", "No active question" } };

            _reactions.Add(reactionTimeMs);
            bool correct = CurrentQuestion.FontColorName.Equals(
                selectedColor, StringComparison.OrdinalIgnoreCase);

            int pointsEarned = 0;

            if (correct)
            {
                Score++;
                Combo++;
                MaxCombo = Math.Max(MaxCombo, Combo);
                pointsEarned = ScoreCalculator.CalculatePoints(reactionTimeMs, Combo, Difficulty);
                TotalPoints += pointsEarned;

                if (Mode == GameMode.Survival) TimeLeft += 3;
                if (Score % 5 == 0) ScaleDifficulty();
            }
            else
            {
                Combo = 0;
                WrongAnswers++;

                switch (Mode)
                {
                    case GameMode.Endless:
                        Lives--;
                        if (Lives <= 0) State = GameState.Finished;
                        break;
                    case GameMode.Survival:
                        TimeLeft = Math.Max(0, TimeLeft - 3);
                        if (TimeLeft <= 0) State = GameState.Finished;
                        break;
                    case GameMode.SpeedRun:
                        TotalPoints = Math.Max(0, TotalPoints - 5);
                        break;
                }
            }

            if (Mode == GameMode.SpeedRun && Score >= Target)
                State = GameState.Finished;

            return new Dictionary<string, object>
            {
                { "correct", correct },
                { "pointsEarned", pointsEarned },
                { "totalPoints", TotalPoints },
                { "score", Score },
                { "combo", Combo },
                { "maxCombo", MaxCombo },
                { "lives", Lives },
                { "timeLeft", Math.Round(TimeLeft, 1) },
                { "isActive", State == GameState.Active }
            };
        }

        private void ScaleDifficulty()
        {
            if (Score >= 40) Difficulty = 5;
            else if (Score >= 30) Difficulty = 4;
            else if (Score >= 20) Difficulty = 3;
            else if (Score >= 10) Difficulty = 2;
            else Difficulty = 1;
            _colorPool = ColorRegistry.GetColorPool(Difficulty);
        }

        /// <summary>Generate the final session report.</summary>
        public SessionResult GetFinalReport()
        {
            double elapsed = (DateTime.Now - StartTime).TotalSeconds;
            double avgRT = _reactions.Count > 0 ? _reactions.Average() : 0;
            int totalAnswers = Score + WrongAnswers;

            var result = new SessionResult
            {
                Mode = Mode,
                TotalPoints = TotalPoints,
                CorrectAnswers = Score,
                WrongAnswers = WrongAnswers,
                MaxCombo = MaxCombo,
                AvgReactionMs = avgRT,
                FastestReaction = _reactions.Count > 0 ? _reactions.Min() : 0,
                SlowestReaction = _reactions.Count > 0 ? _reactions.Max() : 0,
                ElapsedSeconds = elapsed,
                Accuracy = totalAnswers > 0 ? (Score * 100.0 / totalAnswers) : 0,
                CoinsEarned = ScoreCalculator.CalculateCoins(TotalPoints),
                StarsEarned = ScoreCalculator.CalculateStars(Score)
            };
            result.CalculateRating();
            return result;
        }
    }

    // ── Program Entry — Demo ─────────────────────────────────

    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("=== Color Confusion C# Engine ===\n");
            var random = new Random();

            // Test all three modes
            var modes = new[] { GameMode.Endless, GameMode.Survival, GameMode.SpeedRun };

            foreach (var mode in modes)
            {
                Console.WriteLine($"\n--- Testing {mode} Mode ---");
                var session = new GameSession(mode);
                session.Start();

                int rounds = mode == GameMode.SpeedRun ? 15 : 10;

                for (int i = 0; i < rounds; i++)
                {
                    var q = session.NextQuestion();
                    if (q == null) break;

                    bool shouldBeCorrect = random.NextDouble() < 0.7;
                    string answer = shouldBeCorrect ? q.FontColorName : "WrongColor";
                    int reactionTime = 200 + random.Next(1800);

                    var result = session.SubmitAnswer(answer, reactionTime);

                    bool correct = (bool)result["correct"];
                    Console.Write($"  R{i + 1}: {(correct ? "✓" : "✗")} ");
                    Console.Write($"Score:{result["score"]} Points:{result["totalPoints"]} ");
                    Console.Write($"Combo:{result["combo"]} ");

                    if (mode == GameMode.Endless)
                        Console.Write($"Lives:{result["lives"]}");
                    else if (mode == GameMode.Survival)
                        Console.Write($"Time:{result["timeLeft"]}s");

                    Console.WriteLine();

                    if (!(bool)result["isActive"])
                    {
                        Console.WriteLine("  >> GAME OVER");
                        break;
                    }
                }

                // Print report
                var report = session.GetFinalReport();
                Console.WriteLine(report);
            }
        }
    }
}
